### PR TITLE
More integration tests

### PR DESCRIPTION
**Pull request by felipesere**
Adds a couple of integration tests and some plumbing.

### CODE REVIEW

**Comment by felipesere** [comment]
Seems like it fails because the `rust:latest` docker image does not have a real bash...

**Comment by jakedeichert** [comment]
@felipesere it looks like you've taken all my joy away by writing tests 😅 
I really appreciate this though. 

I'll take a look at this later and see what can be done to get the tests to work 👍

**Comment by jakedeichert** [comment]
Looks like the PR diff is out of sync with master for some reason 🤔

Maybe pulling in the latest commits from master will refresh it?

And yeah, looks like the rust image doesn’t like the bash if statement. Maybe try switching to ‘sh’ instead of bash for those code blocks?

**Comment by felipesere** [comment]
Oh yeah I totally need to rebase it.

Gruß,
Felipe

Sent via Superhuman iOS ( https://sprh.mn/?vip=felipesere@gmail.com )

**Comment by felipesere** [comment]
It took a few commits, but now I have it running with bash, python, and node. 
There might be advantages to creating our own rustc container that already includes those at some point... but for now: Its all greeeeen 💚 📗 🍏 !

**Comment by jakedeichert** [comment]
Awesome! I'll take another look late tonight and then squash-merge :100:

File: tests/arguments_and_flags.rs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+"#,
+    );
+
+    common::run_mask(&maskfile_path)
+        .command("test")
+        .arg("the_file")
+        .arg("some_test_case")
+        .assert()
+        .stdout(contains("Testing some_test_case in the_file"))
+        .success();
```
Comment (jakedeichert) [inline tests/arguments_and_flags.rs]: Nice, I really like this cmd testing api :100:

File: tests/subcommands.rs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    common::run_mask(&maskfile_path)
+        .cli("services start my_fancy_service")
+        .assert()
+        .stdout(contains("Starting service my_fancy_service"))
+        .success();
+
+    /*
+     * Doesn't really works.
+     * It triggers the 'service stop $name
+     *
```
Comment (jakedeichert) [inline tests/subcommands.rs]: Yupp, i think this makes sense. For now I don't see a real need to support command names _after_ required/optional args. The workaround is just to make `now` another required arg like `(when)` for example.

I feel like we can probably delete this commented out block and also delete the ` #### services stop (service_name) now` section from the maskfile above 👍

File: tests/supported_runtimes_test.rs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+use predicates::str::contains;
+
+mod common;
+use common::MaskCommandExt;
+
+#[test]
+fn bash() {
+    let (_temp, maskfile_path) = common::maskfile(
+        "
+# Integartion tests
```
Comment (jakedeichert) [inline tests/supported_runtimes_test.rs]:
```suggestion
# Integration tests
```

File: tests/supported_runtimes_test.rs [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+        .assert()
+        .stdout(contains("Hello, World!"))
+        .success();
+}
+
+#[test]
+fn node() {
+    let (_temp, maskfile_path) = common::maskfile(
+        "
+# Integartion tests
```
Comment (jakedeichert) [inline tests/supported_runtimes_test.rs]:
```suggestion
# Integration tests
```

File: tests/supported_runtimes_test.rs [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        .assert()
+        .stdout(contains("Hello, World!"))
+        .success();
+}
+
+#[test]
+fn python() {
+    let (_temp, maskfile_path) = common::maskfile(
+        r#"
+# Integartion tests
```
Comment (jakedeichert) [inline tests/supported_runtimes_test.rs]:
```suggestion
# Integration tests
```

File: tests/supported_runtimes_test.rs [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+        .assert()
+        .stdout(contains("Hello, World!"))
+        .success();
+}
+
+#[test]
+fn ruby() {
+    let (_temp, maskfile_path) = common::maskfile(
+        r#"
+# Integartion tests
```
Comment (jakedeichert) [inline tests/supported_runtimes_test.rs]:
```suggestion
# Integration tests
```

File: tests/supported_runtimes_test.rs [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+        .assert()
+        .stdout(contains("Hello, World!"))
+        .success();
+}
+
+#[test]
+fn php() {
+    let (_temp, maskfile_path) = common::maskfile(
+        r#"
+# Integartion tests
```
Comment (jakedeichert) [inline tests/supported_runtimes_test.rs]:
```suggestion
# Integration tests
```

File: tests/supported_runtimes_test.rs [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+
+",
+    );
+
+    common::run_mask(&maskfile_path)
+        .command("bash")
+        .env("name", "World")
+        .assert()
+        .stdout(contains("Hello, World!"))
+        .success();
```
Comment (jakedeichert) [inline tests/supported_runtimes_test.rs]: Nice! Really happy each runtime is now tested.

File: tests/subcommands.rs (hunk 2 above)
Comment (felipesere) [reply tests/subcommands.rs]: 👍  I'll remove it from here and the `README.md`, which is where I got it from.

**Comment by felipesere** [comment]
Should be all up to date now 😸

**Pull request merged**